sdk/src/main/java/io/opentelemetry/sdk/metrics/BoundRegistry.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright 2020, OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.opentelemetry.sdk.metrics;
+
+import io.opentelemetry.metrics.LabelSet;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.locks.ReentrantLock;
+
+abstract class BoundRegistry<B extends AbstractBoundInstrument> {
+  private final ConcurrentHashMap<LabelSet, B> bounds;
+  private final ReentrantLock collectLock;
+
+  BoundRegistry() {
+    bounds = new ConcurrentHashMap<>();
+    collectLock = new ReentrantLock();
+  }
+
+  B acquire(LabelSet labelSet) {
+    B bound = bounds.get(labelSet);
+    if (bound != null && bound.ref()) {
+      // At this moment it is guaranteed that the Bound is in the map and will not be removed.
+      return bound;
+    }
+
+    // Missing entry or no longer mapped, try to add a new entry.
+    bound = newBound(labelSet);
+    while (true) {
+      B oldBound = bounds.putIfAbsent(labelSet, bound);
+      if (oldBound != null) {
+        if (oldBound.ref()) {
+          // At this moment it is guaranteed that the Bound is in the map and will not be removed.
+          return oldBound;
+        }
+        // Try to remove the oldBound. This will race with the collect method, but only one will
+        // succeed.
+        bounds.remove(labelSet, oldBound);
+        continue;
+      }
+      return bound;
+    }
+  }
+
+  void release(B bound) {
+    bound.unref();
+  }
+
+  /**
+   * Collects records from all the entries (labelSet, Bound) that changed since the last collect()
+   * call.
+   *
+   * <p>It is possible that some entries
+   */
+  void collect(RecordProcessor recordProcessor) {
+    collectLock.lock();
```

Comment:
I really think we need to find a lockless way of handling this, as this will be a major source of contention in high-throughput situations.

Reply:
Collect is called async, the only reason to have the lock is to ensure one collect call at any moment.

Reply:
fair enough.